File: lab_4/zad_1/prime_functions.c
#include "prime_functions.h"

#include <math.h>
#include <stdio.h>
#include <stdbool.h>


void fill_sieve(bool sieve[], unsigned n) {
    sieve[0] = false;
    sieve[1] = false;
    for (int i = 2; i <= n; ++i) { sieve[i] = true; }
}


void calculate_sieve(bool sieve[], unsigned n) {
    unsigned sqrt_value = (unsigned) sqrt(n);

    for (unsigned i = 2; i <= sqrt_value; ++i) {
        if (sieve[i]) {
            for (unsigned j = i * i; j < n; j += i) {
                sieve[j] = false;
            }
        }
    }
}


unsigned count_primes(bool sieve[], unsigned n) {
    int count = 0;

    for (int i = 0; i < n; ++i) {
        if (sieve[i]) {count++;}
    }

    return count;
}

unsigned prime_numbers(unsigned n) {
    bool sieve[n + 1];

    fill_sieve(sieve, n);
    calculate_sieve(sieve, n);
    unsigned count = count_primes(sieve, n);

    return count;
}

bool is_prime(unsigned n) {
    // used from zad_2 z lab_2
    if (n <= 1) return false;
    if (n == 2) return true;
    if (n % 2 == 0) return false;

    float sqrt_n = sqrt(n);
    for (int i = 3; i < sqrt_n; i += 2) {
        if (n % i == 0) {
            return false;
        }
    }
    return true;
}

unsigned prime(unsigned n) {
    int upper_limit;

    if (n < 5) {
        upper_limit = (n * 2) + 1;
    }
    else {
        upper_limit = n + (n * (int)(log(n) + log(log(n))));
    }

    bool sieve[upper_limit];

    fill_sieve(sieve, upper_limit);
    calculate_sieve(sieve, upper_limit);

    unsigned i = 0, count = 0;
    while (count < n) {
        if (sieve[i]) {
            count++;
        }
        i++;
    }
    unsigned result = i-1;

    return result;
}

File: lab_4/zad_3/prime_functions.py
import math

def fill_sieve(n):
    sieve = [True] * (n + 1)
    sieve[0] = sieve[1] = False
    return sieve

def PrimeNumbers(n):
    sieve = fill_sieve(n)
    limit = int(math.sqrt(n))
    for i in range(2, limit + 1):
        if sieve[i]:
            for j in range(i * i, n + 1, i):
                sieve[j] = False
    
    return sum(sieve)

def Prime(n):
    if n < 5:
        upper_limit = n * 2 + 1
    else:
        upper_limit = n + int(n * (math.log(n) + math.log(math.log(n))))

    sieve = [False] * (upper_limit)
    
    fill_sieve(sieve, upper_limit)
    calculate_sieve(sieve, upper_limit)

    count = 0
    for i, number in enumerate(sieve):
        if number:
            count += 1
            if count == n:
                return i

def IsPrime(n):
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    return True
